src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpContentReadStream.cs

```diff
@@ -38,5 +38,41 @@ public sealed override Task<int> ReadAsync(byte[] buffer, int offset, int count,
 
         public sealed override void CopyTo(Stream destination, int bufferSize) =>
             CopyToAsync(destination, bufferSize, CancellationToken.None).GetAwaiter().GetResult();
+
+        public virtual Task DrainAsync(int maxDrainBytes) => Task.CompletedTask;
+
+        protected override void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                if (_connection != null)
+                {
+                    // Start the asynchronous drain.
+                    // It may complete synchronously, in which case the connection will be put back in the pool synchronously.
+                    // Skip the call to base.Dispose -- it will be deferred until DrainOnDispose finishes.
+                    DrainOnDispose();
+                    return;
+                }
+            }
+
+            base.Dispose(disposing);
+        }
+
+        // Maximum request drain size, 16K.
+        private const int s_maxDrainBytes = 16 * 1024;
+
+        private async void DrainOnDispose()
+        {
+            try
+            {
+                await DrainAsync(s_maxDrainBytes);
+            }
+            catch (Exception)
+            {
+                // Eat any exceptions and just Dispose.
+            }
+
+            base.Dispose(true);
```

Comment:
We Dispose even if we successfully drained everything?  Doesn't that mean we may end up disposing a connection that's already been put back into the pool?

Reply:
When the connection is put back in the pool, we set _connection to null.  So base.Dispose will not dispose it again.

That said, currently base.Dispose doesn't do anything else except dispose the connection (if not null).  So it's not technically necessary, but still seems like the right thing to do.